runtime/compiler/runtime/JITServerAOTCache.cpp

```diff
@@ -745,6 +1091,357 @@ JITServerAOTCache::printStats(FILE *f) const
+      }
+      {
+      OMR::CriticalSection cs(_wellKnownClassesMonitor);
+      header._numWellKnownClassesRecords = _wellKnownClassesMap.size();
+      header._nextWellKnownClassesId = _nextWellKnownClassesId;
+      }
+      {
+      OMR::CriticalSection cs(_classChainMonitor);
+      header._numClassChainRecords = _classChainMap.size();
+      header._nextClassChainId = _nextClassChainId;
+      }
+      {
+      OMR::CriticalSection cs(_methodMonitor);
+      header._numMethodRecords = _methodMap.size();
+      header._nextMethodId = _nextMethodId;
+      }
+      {
+      OMR::CriticalSection cs(_classMonitor);
+      header._numClassRecords = _classMap.size();
+      header._nextClassId = _nextClassId;
+      }
+      {
+      OMR::CriticalSection cs(_classLoaderMonitor);
+      header._numClassLoaderRecords = _classLoaderMap.size();
+      header._nextClassLoaderId = _nextClassLoaderId;
+      }
+
+   if (1 != fwrite(&header, sizeof(JITServerAOTCacheHeader), 1, f))
+      return false;
+
+   if (!writeAOTRecordList(f, _classLoaderHead, header._numClassLoaderRecords))
+      return false;
+   if (!writeAOTRecordList(f, _classHead, header._numClassRecords))
+      return false;
+   if (!writeAOTRecordList(f, _methodHead, header._numMethodRecords))
+      return false;
+   if (!writeAOTRecordList(f, _classChainHead, header._numClassChainRecords))
+      return false;
+   if (!writeAOTRecordList(f, _wellKnownClassesHead, header._numWellKnownClassesRecords))
+      return false;
+   if (!writeAOTRecordList(f, _aotHeaderHead, header._numAOTHeaderRecords))
+      return false;
+   if (!writeCachedMethodList(f, _cachedMethodHead, header._numCachedAOTMethods))
+      return false;
+
+   return true;
+   }
+
+// Tests whether or not the given AOT snapshot is compatible with the server.
+static bool
+isCompatibleSnapshotVersion(const JITServerAOTCacheVersion &version)
+   {
+   JITServerAOTCacheVersion currentVersion = {0};
+   getCurrentAOTCacheVersion(currentVersion);
+
+   return (0 == memcmp(version._eyeCatcher, currentVersion._eyeCatcher, sizeof(currentVersion._eyeCatcher))) &&
+          (version._snapshotVersion == currentVersion._snapshotVersion) &&
+          (version._jitserverVersion == currentVersion._jitserverVersion);
+   }
+
+// Read an AOT cache snapshot, returning NULL if the cache is ill-formed or
+// incompatible with the running server.
+JITServerAOTCache *
+JITServerAOTCache::readCache(FILE *f, const std::string &name, TR_Memory &trMemory)
+   {
+   if (!JITServerAOTCacheMap::cacheHasSpace())
+      return NULL;
+
+   JITServerAOTCacheHeader header = {0};
+   if (1 != fread(&header, sizeof(JITServerAOTCacheHeader), 1, f))
+      return NULL;
+
+   if (!isCompatibleSnapshotVersion(header._version))
+      return NULL;
+
+   JITServerAOTCache *cache = NULL;
+   try
+      {
+      cache = new (TR::Compiler->persistentGlobalMemory()) JITServerAOTCache(name);
+      }
+   catch (const std::bad_alloc& e)
```

Comment:
I think it would be better to catch std:: exception, and also write a message to vlog including the exception message.